test/infrastructure/container/docker.go

```diff
@@ -121,55 +126,212 @@ func (d *docker) GetHostPort(ctx context.Context, name, portAndProtocol string)
-func (d *docker) ExecToFile(ctx context.Context, containerName string, fileOnHost *os.File, command string, args ...string) error {
+// Exec executes a command in a running container and writes any output to the provided writer.
+func (d *docker) Exec(ctxd context.Context, containerName string, config *ExecContainerInput, command string, args ...string) error {
+	ctx := context.Background() // Let the command finish, even if it takes longer than the default timeout
 	execConfig := types.ExecConfig{
+		// Run with privileges so we can remount etc..
+		// This might not make sense in the most general sense, but it is
+		// important to many kind commands.
 		Privileged:   true,
 		Cmd:          append([]string{command}, args...),
 		AttachStdout: true,
 		AttachStderr: true,
+		AttachStdin:  config.InputBuffer != nil,
+		Env:          config.EnvironmentVars,
 	}
 
 	response, err := d.dockerClient.ContainerExecCreate(ctx, containerName, execConfig)
 	if err != nil {
-		return fmt.Errorf("error creating container exec: %v", err)
+		return errors.Wrap(err, "error creating container exec")
 	}
 
 	execID := response.ID
 	if execID == "" {
-		return fmt.Errorf("exec ID empty")
+		return errors.Wrap(err, "exec ID empty")
 	}
 
 	resp, err := d.dockerClient.ContainerExecAttach(ctx, execID, types.ExecStartCheck{})
 	if err != nil {
-		return fmt.Errorf("error attaching to container exec: %v", err)
+		return errors.Wrap(err, "error attaching to container exec")
 	}
 	defer resp.Close()
 
-	// Read out any output from the call
+	// If there is input, send it through to its stdin
+	inputErrors := make(chan error)
+	if config.InputBuffer != nil {
+		go func() {
+			_, err := io.Copy(resp.Conn, config.InputBuffer)
+			inputErrors <- err
+			_ = resp.CloseWrite()
+		}()
+	}
+
+	if config.OutputBuffer == nil {
+		// We always want to read whatever output the command sends
+		config.OutputBuffer = &bytes.Buffer{}
+	}
+
 	outputErrors := make(chan error)
 	go func() {
-		// Send the output to the host file
-		_, err = stdcopy.StdCopy(fileOnHost, fileOnHost, resp.Reader)
+		// Send the output to the output writer
+		var err error
+		if config.ErrorBuffer != nil {
+			_, err = stdcopy.StdCopy(config.OutputBuffer, config.ErrorBuffer, resp.Reader)
+		} else {
+			_, err = io.Copy(config.OutputBuffer, resp.Reader)
+		}
 		outputErrors <- err
+		close(outputErrors)
 	}()
 
 	select {
+	case err := <-inputErrors:
+		if err != nil {
+			return errors.Wrap(err, "error providing execution input")
+		}
+
 	case err := <-outputErrors:
 		if err != nil {
-			return fmt.Errorf("error reading output from container exec: %v", err)
+			return errors.Wrap(err, "error getting execution output")
 		}
 
 	case <-ctx.Done():
-		return err
+		return errors.Wrap(ctx.Err(), "operation cancelled")
+	}
+
+	// Explicitly close the connection since we are done reading so it can
+	// exit.
+	resp.Close()
+
+	retry := 0
+	for retry < 600 {
```

Comment:
I know this is mostly due to the docker API, but is there a better way to wait for command completion (instead of retrying)

Reply:
Not that I have been able to find. At least no helpers in the API itself.

This may be less necessary to wait on now though. Originally I was not setting the output buffer on line 173 if none was provided. That appeared to be the root of an issue I was seeing where commands were not completing completely. Now that it is always set and there is the wait for the output to be read for every command, the original issue I was seeing appears to be avoided.

I will do some testing with this loop removed. Not sure if we want to hold up anything on that since from what I've seen in local tests it's usually at most one retry. But I will try to clean this up if it doesn't look like it is really needed

Reply:
I did some more local tests. It doesn't look like this is necessary anymore, so I've dropped it for now.

Reply:
And added it back. Not sure exactly what happens, but my assumption is some of the longer commands (like kubeadm join) don't quite finish right away, then cause issues.